database/src/ops/block.rs

```diff
@@ -1,89 +1,321 @@
+        AmountIndex, BlockHash, BlockHeight, BlockInfoLatest, BlockInfoV1, BlockInfoV2,
+        BlockInfoV3, KeyImage, Output, PreRctOutputId, RctOutput, TxHash,
+    },
+};
 
-/// TODO
-pub fn pop_block() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps and adds them to the appropriate database tables.
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[allow(clippy::too_many_lines)]
+// no inline, too big.
+pub fn add_block(
+    tables: &mut impl TablesMut,
+    block: &VerifiedBlockInformation,
+) -> Result<(), RuntimeError> {
+    // Block Info.
+    //
+    // Branch on the hard fork version (`major_version`)
+    // and add the block to the appropriate table.
+    // <https://monero-book.cuprate.org/consensus_rules/hardforks.html#Mainnet-Hard-Forks>
+    //
+    // FIXME: use `match` with ranges when stable:
+    // <https://github.com/rust-lang/rust/issues/37854>
+    if block.block.header.major_version < 4 {
+        tables.block_info_v1s_mut().put(
+            &block.height,
+            &BlockInfoV1 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                #[allow(clippy::cast_possible_truncation)] // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+            },
+        )
+    } else if block.block.header.major_version < 10 {
+        tables.block_info_v2s_mut().put(
+            &block.height,
+            #[allow(clippy::cast_possible_truncation)] // TODO
+            &BlockInfoV2 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                #[allow(clippy::cast_possible_truncation)] // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(), // TODO
+            },
+        )
+    } else {
+        tables.block_info_v3s_mut().put(
+            &block.height,
+            &BlockInfoV3 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty,
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(),                    // TODO
+                long_term_weight: block.long_term_weight as u64, // TODO
+            },
+        )
+    }?;
```

Comment:
What to do with the size difference between `VerifiedBlockInformation` and our `BlockInfoV*` types?

Casting here is lossy?